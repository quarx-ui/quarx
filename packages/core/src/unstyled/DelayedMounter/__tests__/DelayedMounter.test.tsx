import React, { useState } from 'react';
import { render, screen } from '@testing-library/react';
import { DelayedMounter } from '@core';
import userEvent from '@testing-library/user-event';

describe('DelayedMounter behavior', () => {
    it('text should be in the document', () => {
        render(<DelayedMounter mounted>DelayedMounter</DelayedMounter>);

        expect(screen.queryByText('DelayedMounter')).toBeInTheDocument();
    });

    it('text should not be in the document', () => {
        render(<DelayedMounter mounted={false}>DelayedMounter</DelayedMounter>);

        expect(screen.queryByText('DelayedMounter')).not.toBeInTheDocument();
    });

    it('default layout', () => {
        const { asFragment } = render(
            <DelayedMounter mounted>
                <div>
                    DelayedMounter
                </div>
            </DelayedMounter>,
        );

        expect(asFragment()).toMatchSnapshot();
    });
});

jest.useFakeTimers();
jest.spyOn(global, 'setTimeout');

describe('DelayedMounter timeout', () => {
    const BUTTON_TEXT = 'Mount Component';

    const Template = () => {
        const [mounted, setMounted] = useState(false);

        return (
            <div>
                <DelayedMounter mounted={mounted}>
                    <div>
                        DelayedMounter
                    </div>
                </DelayedMounter>
                <button
                    onClick={() => setMounted((prevState) => !prevState)}
                    type="button"
                >
                    {BUTTON_TEXT}
                </button>
            </div>
        );
    };

    it('layout', async () => {
        const { asFragment } = render(<Template />);
        const trigger = screen.getByText(BUTTON_TEXT);

        // Разметка до клика (компонент отсутствует)
        expect(asFragment()).toMatchSnapshot();

        // Разметка после клика (компонент присутствует)
        userEvent.click(trigger);
        expect(asFragment()).toMatchSnapshot();

        // Разметка после второго клика (компонент еще присутствует)
        userEvent.click(trigger);
        expect(asFragment()).toMatchSnapshot();

        jest.runAllTimers();

        // Разметка после отработки дефолтного таймера (компонент отсутствует)
        expect(asFragment()).toMatchSnapshot();
    });

    it('functions', () => {
        render(<Template />);

        expect(setTimeout).toHaveBeenCalledTimes(0);

        const trigger = screen.getByText(BUTTON_TEXT);

        userEvent.click(trigger);
        userEvent.click(trigger);

        jest.runAllTimers();

        expect(setTimeout).toHaveBeenCalledTimes(1);
        expect(setTimeout).toHaveBeenCalledWith(expect.any(Function), 250);
    });
});
