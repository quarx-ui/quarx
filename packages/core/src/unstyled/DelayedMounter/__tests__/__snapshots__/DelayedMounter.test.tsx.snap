// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DelayedMounter behavior default layout 1`] = `
<DocumentFragment>
  <div>
    DelayedMounter
  </div>
</DocumentFragment>
`;

exports[`DelayedMounter timeout layout 1`] = `
<DocumentFragment>
  <div>
    <button
      type="button"
    >
      Mount Component
    </button>
  </div>
</DocumentFragment>
`;

exports[`DelayedMounter timeout layout 2`] = `
<DocumentFragment>
  <div>
    <div>
      DelayedMounter
    </div>
    <button
      type="button"
    >
      Mount Component
    </button>
  </div>
</DocumentFragment>
`;

exports[`DelayedMounter timeout layout 3`] = `
<DocumentFragment>
  <div>
    <div>
      DelayedMounter
    </div>
    <button
      type="button"
    >
      Mount Component
    </button>
  </div>
</DocumentFragment>
`;

exports[`DelayedMounter timeout layout 4`] = `
<DocumentFragment>
  <div>
    <button
      type="button"
    >
      Mount Component
    </button>
  </div>
</DocumentFragment>
`;
